File: mcp-server/packages/context-extractor-python/src/context_extractor.py
"""
Context Extractor - Extracts method trees from objects using reflection.

This module provides functionality to analyze Python objects and extract
their method hierarchies for use as AI context enrichment.
"""

import inspect
import os
import time
import json
import logging
from typing import Any, Set, Optional, Type, Callable

try:
    from .models import MethodInfo, ParameterInfo, PropertyInfo, ObjectNode
except ImportError:
    from models import MethodInfo, ParameterInfo, PropertyInfo, ObjectNode

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)


class ContextExtractor:
    """
    Extracts method trees and object hierarchies using Python reflection.
    
    This class analyzes objects and their attributes recursively, building
    a structured representation of available methods that can be used to
    enrich AI context for better code completion suggestions.
    """
    
    def __init__(self, output_dir: str = "generated_context", max_depth: int = 10):
        """
        Initialize the ContextExtractor.
        
        Args:
            output_dir: Directory where generated context files will be saved
            max_depth: Maximum depth for recursive object exploration
        """
        self.output_dir = output_dir
        self.max_depth = max_depth
        self.logger = logging.getLogger(self.__class__.__name__)
        
        # Create output directory if it doesn't exist
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
            self.logger.info(f"Created output directory: {output_dir}")
    
    def extract_methods(self, obj: Any, include_private: bool = False) -> list[MethodInfo]:
        """
        Extract all methods from an object.
        
        Args:
            obj: The object to extract methods from
            include_private: Whether to include methods starting with underscore
            
        Returns:
            List of MethodInfo objects describing the methods
        """
        methods = []
        
        for name, member in inspect.getmembers(obj):
            # Skip magic methods and optionally private methods
            if name.startswith("__"):
                continue
            if not include_private and name.startswith("_"):
                continue
            
            # Check if it's a method
            if inspect.ismethod(member) or inspect.isfunction(member):
                method_info = self._analyze_method(name, member)
                if method_info:
                    methods.append(method_info)
        
        return methods
    
    def _analyze_method(self, name: str, method: Callable) -> Optional[MethodInfo]:
        """
        Analyze a method and extract its signature information.
        
        Args:
            name: The method name
            method: The method object
            
        Returns:
            MethodInfo object or None if analysis fails
        """
        try:
            sig = inspect.signature(method)
            
            # Extract parameters
            parameters = []
            for param_name, param in sig.parameters.items():
                if param_name == 'self':
                    continue
                
                # Get parameter type
                type_name = "Any"
                if param.annotation != inspect.Parameter.empty:
                    type_name = self._get_type_name(param.annotation)
                
                # Get default value
                default_value = None
                if param.default != inspect.Parameter.empty:
                    default_value = repr(param.default)
                
                parameters.append(ParameterInfo(
                    name=param_name,
                    type_name=type_name,
                    default_value=default_value
                ))
            
            # Get return type
            return_type = "Any"
            if sig.return_annotation != inspect.Signature.empty:
                return_type = self._get_type_name(sig.return_annotation)
            
            # Get docstring
            docstring = inspect.getdoc(method)
            
            # Check if static or class method
            is_static = isinstance(inspect.getattr_static(method.__self__.__class__, name, None), staticmethod) if hasattr(method, '__self__') else False
            is_class_method = isinstance(inspect.getattr_static(method.__self__.__class__, name, None), classmethod) if hasattr(method, '__self__') else False
            
            return MethodInfo(
                name=name,
                parameters=parameters,
                return_type=return_type,
                docstring=docstring,
                is_static=is_static,
                is_class_method=is_class_method
            )
            
        except Exception as e:
            self.logger.debug(f"Could not analyze method {name}: {e}")
            return MethodInfo(
                name=name,
                parameters=[],
                return_type="Unknown",
                docstring="Could not inspect method"
            )
    
    def _get_type_name(self, annotation: Any) -> str:
        """
        Get a string representation of a type annotation.
        
        Args:
            annotation: The type annotation
            
        Returns:
            String representation of the type
        """
        if hasattr(annotation, '__name__'):
            return annotation.__name__
        elif hasattr(annotation, '__origin__'):
            # Handle generic types like List[str], Dict[str, int], etc.
            origin = annotation.__origin__
            args = getattr(annotation, '__args__', ())
            origin_name = getattr(origin, '__name__', str(origin))
            if args:
                args_str = ", ".join(self._get_type_name(a) for a in args)
                return f"{origin_name}[{args_str}]"
            return origin_name
        return str(annotation)
    
    def extract_properties(self, obj: Any, include_private: bool = False) -> list[PropertyInfo]:
        """
        Extract properties/attributes from an object.
        
        Args:
            obj: The object to extract properties from
            include_private: Whether to include private attributes
            
        Returns:
            List of PropertyInfo objects
        """
        properties = []
        
        for name in dir(obj):
            if name.startswith("__"):
                continue
            if not include_private and name.startswith("_"):
                continue
            
            try:
                value = getattr(obj, name)
                # Skip methods
                if callable(value):
                    continue
                
                type_name = type(value).__name__
                properties.append(PropertyInfo(
                    name=name,
                    type_name=type_name,
                    is_readonly=False  # Python doesn't have true read-only properties by default
                ))
            except AttributeError:
                continue
        
        return properties
    
    def explore_object(self, obj: Any, name: str = "root", 
                       visited: Optional[Set[int]] = None, 
                       depth: int = 0,
                       include_private: bool = False) -> ObjectNode:
        """
        Recursively explore an object and build its hierarchy tree.
        
        Args:
            obj: The object to explore
            name: Name to assign to this node
            visited: Set of already visited object IDs (for cycle detection)
            depth: Current depth in the hierarchy
            include_private: Whether to include private members
            
        Returns:
            ObjectNode representing the object and its hierarchy
        """
        if visited is None:
            visited = set()
        
        obj_id = id(obj)
        class_name = obj.__class__.__name__
        
        # Create the node
        node = ObjectNode(
            name=name,
            class_name=class_name,
            depth=depth
        )
        
        # Check for cycles or max depth
        if obj_id in visited or depth >= self.max_depth:
            return node
        
        visited.add(obj_id)
        
        # Extract methods
        node.methods = self.extract_methods(obj, include_private)
        
        # Extract properties
        node.properties = self.extract_properties(obj, include_private)
        
        # Explore sub-objects (attributes that are custom objects)
        for attr_name, attr_value in vars(obj).items():
            if attr_name.startswith("__"):
                continue
            if not include_private and attr_name.startswith("_"):
                continue
            
            # Only explore custom objects, not primitives
            if self._is_explorable_object(attr_value):
                child_node = self.explore_object(
                    attr_value, 
                    attr_name, 
                    visited.copy(), 
                    depth + 1,
                    include_private
                )
                node.children.append(child_node)
        
        return node
    
    def _is_explorable_object(self, obj: Any) -> bool:
        """
        Determine if an object should be explored recursively.
        
        Args:
            obj: The object to check
            
        Returns:
            True if the object should be explored
        """
        # Skip None and primitive types
        if obj is None:
            return False
        
        primitive_types = (str, int, float, bool, bytes, type(None))
        if isinstance(obj, primitive_types):
            return False
        
        # Skip common container types (unless they contain custom objects)
        container_types = (list, dict, set, tuple, frozenset)
        if isinstance(obj, container_types):
            return False
        
        # Skip types themselves
        if isinstance(obj, type):
            return False
        
        # Check if it has custom attributes (indicates a custom class)
        return hasattr(obj, '__dict__') and len(vars(obj)) > 0
    
    def save_as_json(self, obj: Any, filename: Optional[str] = None, 
                     include_private: bool = False) -> str:
        """
        Extract object hierarchy and save as JSON.
        
        Args:
            obj: The root object to analyze
            filename: Output filename (auto-generated if None)
            include_private: Whether to include private members
            
        Returns:
            Path to the saved file
        """
        if filename is None:
            timestamp = int(time.time())
            class_name = obj.__class__.__name__.lower()
            filename = f"{class_name}_{timestamp}.json"
        
        filepath = os.path.join(self.output_dir, filename)
        
        # Extract the object tree
        root_node = self.explore_object(obj, include_private=include_private)
        
        # Save to file
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(root_node.to_json())
        
        self.logger.info(f"Saved JSON context to: {filepath}")
        return filepath
    
    def save_as_text(self, obj: Any, filename: Optional[str] = None,
                     include_private: bool = False,
                     include_details: bool = True) -> str:
        """
        Extract object hierarchy and save as AI-optimized text.
        
        Args:
            obj: The root object to analyze
            filename: Output filename (auto-generated if None)
            include_private: Whether to include private members
            include_details: Whether to include detailed parameter info
            
        Returns:
            Path to the saved file
        """
        if filename is None:
            timestamp = int(time.time())
            class_name = obj.__class__.__name__.lower()
            filename = f"{class_name}_{timestamp}.txt"
        
        filepath = os.path.join(self.output_dir, filename)
        
        # Extract the object tree
        root_node = self.explore_object(obj, include_private=include_private)
        
        # Save to file
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(root_node.to_text(include_details))
        
        self.logger.info(f"Saved text context to: {filepath}")
        return filepath
    
    def save(self, result: "ObjectNode", filename: str, format: str = "json") -> str:
        """
        Save extraction result to a file.
        
        Args:
            result: The ObjectNode to save
            filename: Base filename (without extension)
            format: Output format - 'json' or 'text'
            
        Returns:
            Path to the saved file
        """
        if format == "json":
            full_filename = f"{filename}.json"
            filepath = os.path.join(self.output_dir, full_filename)
            with open(filepath, 'w', encoding='utf-8') as f:
                f.write(result.to_json())
        else:
            full_filename = f"{filename}.txt"
            filepath = os.path.join(self.output_dir, full_filename)
            with open(filepath, 'w', encoding='utf-8') as f:
                f.write(result.to_text())
        
        self.logger.info(f"Saved context to: {filepath}")
        return filepath
    
    def extract_from_class(self, cls: Type, include_private: bool = False) -> ObjectNode:
        """
        Extract method information from a class (without instantiating).
        
        Args:
            cls: The class to analyze
            include_private: Whether to include private members
            
        Returns:
            ObjectNode representing the class structure
        """
        node = ObjectNode(
            name=cls.__name__,
            class_name=cls.__name__,
            depth=0
        )
        
        for name, member in inspect.getmembers(cls):
            if name.startswith("__"):
                continue
            if not include_private and name.startswith("_"):
                continue
            
            if inspect.isfunction(member) or inspect.ismethod(member):
                try:
                    sig = inspect.signature(member)
                    
                    parameters = []
                    for param_name, param in sig.parameters.items():
                        if param_name == 'self':
                            continue
                        
                        type_name = "Any"
                        if param.annotation != inspect.Parameter.empty:
                            type_name = self._get_type_name(param.annotation)
                        
                        default_value = None
                        if param.default != inspect.Parameter.empty:
                            default_value = repr(param.default)
                        
                        parameters.append(ParameterInfo(
                            name=param_name,
                            type_name=type_name,
                            default_value=default_value
                        ))
                    
                    return_type = "Any"
                    if sig.return_annotation != inspect.Signature.empty:
                        return_type = self._get_type_name(sig.return_annotation)
                    
                    node.methods.append(MethodInfo(
                        name=name,
                        parameters=parameters,
                        return_type=return_type,
                        docstring=inspect.getdoc(member)
                    ))
                except (ValueError, TypeError):
                    node.methods.append(MethodInfo(
                        name=name,
                        parameters=[],
                        return_type="Unknown",
                        docstring="Could not inspect"
                    ))
        
        return node

File: mcp-server/packages/context-extractor-java/src/main/java/io/contextextractor/ContextExtractor.java
package io.contextextractor;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.lang.reflect.Parameter;
import java.util.*;
import java.util.logging.Logger;

/**
 * Context Extractor - Extracts method trees from objects using Java Reflection.
 * 
 * This class analyzes objects and their fields recursively, building
 * a structured representation of available methods that can be used to
 * enrich AI context for better code completion suggestions.
 */
public class ContextExtractor {
    private static final Logger logger = Logger.getLogger(ContextExtractor.class.getName());
    
    private final String outputDir;
    private final int maxDepth;
    private final boolean includePrivate;

    /**
     * Create a ContextExtractor with default settings.
     */
    public ContextExtractor() {
        this("generated_context", 10, false);
    }

    /**
     * Create a ContextExtractor with custom settings.
     * 
     * @param outputDir Directory for generated files
     * @param maxDepth Maximum depth for recursive exploration
     * @param includePrivate Whether to include private methods
     */
    public ContextExtractor(String outputDir, int maxDepth, boolean includePrivate) {
        this.outputDir = outputDir;
        this.maxDepth = maxDepth;
        this.includePrivate = includePrivate;
        
        // Create output directory if it doesn't exist
        File dir = new File(outputDir);
        if (!dir.exists()) {
            dir.mkdirs();
            logger.info("Created output directory: " + outputDir);
        }
    }

    /**
     * Extract methods from a class.
     * 
     * @param clazz The class to analyze
     * @return List of MethodInfo objects
     */
    public List<MethodInfo> extractMethods(Class<?> clazz) {
        List<MethodInfo> methods = new ArrayList<>();
        
        for (Method method : clazz.getDeclaredMethods()) {
            // Skip synthetic and bridge methods
            if (method.isSynthetic() || method.isBridge()) {
                continue;
            }
            
            // Check visibility
            int modifiers = method.getModifiers();
            if (!includePrivate && !Modifier.isPublic(modifiers)) {
                continue;
            }
            
            MethodInfo methodInfo = analyzeMethod(method);
            methods.add(methodInfo);
        }
        
        // Sort methods by name for consistent output
        methods.sort(Comparator.comparing(MethodInfo::getName));
        
        return methods;
    }

    /**
     * Extract methods from an object instance.
     * 
     * @param obj The object to analyze
     * @return List of MethodInfo objects
     */
    public List<MethodInfo> extractMethods(Object obj) {
        return extractMethods(obj.getClass());
    }

    /**
     * Analyze a single method and extract its information.
     */
    private MethodInfo analyzeMethod(Method method) {
        String name = method.getName();
        
        // Extract parameters
        List<ParameterInfo> parameters = new ArrayList<>();
        for (Parameter param : method.getParameters()) {
            String paramName = param.getName();
            String typeName = getTypeName(param.getType());
            parameters.add(new ParameterInfo(paramName, typeName));
        }
        
        // Get return type
        String returnType = getTypeName(method.getReturnType());
        
        // Check modifiers
        int modifiers = method.getModifiers();
        boolean isStatic = Modifier.isStatic(modifiers);
        boolean isPublic = Modifier.isPublic(modifiers);
        
        return MethodInfo.builder(name)
                .parameters(parameters)
                .returnType(returnType)
                .isStatic(isStatic)
                .isPublic(isPublic)
                .build();
    }

    /**
     * Get a clean type name for a class.
     */
    private String getTypeName(Class<?> type) {
        if (type.isArray()) {
            return getTypeName(type.getComponentType()) + "[]";
        }
        
        String name = type.getSimpleName();
        if (name.isEmpty()) {
            name = type.getName();
            int lastDot = name.lastIndexOf('.');
            if (lastDot >= 0) {
                name = name.substring(lastDot + 1);
            }
        }
        return name;
    }

    /**
     * Explore an object and build its hierarchy tree.
     * 
     * @param obj The object to explore
     * @return ObjectNode representing the hierarchy
     */
    public ObjectNode exploreObject(Object obj) {
        return exploreObject(obj, "root", new HashSet<>(), 0);
    }

    /**
     * Explore an object with a custom root name.
     * 
     * @param obj The object to explore
     * @param name Name for the root node
     * @return ObjectNode representing the hierarchy
     */
    public ObjectNode exploreObject(Object obj, String name) {
        return exploreObject(obj, name, new HashSet<>(), 0);
    }

    /**
     * Recursively explore an object.
     */
    private ObjectNode exploreObject(Object obj, String name, Set<Integer> visited, int depth) {
        Class<?> clazz = obj.getClass();
        String className = clazz.getSimpleName();
        
        ObjectNode.Builder nodeBuilder = ObjectNode.builder(name, className)
                .depth(depth);
        
        // Check for cycles or max depth
        int objId = System.identityHashCode(obj);
        if (visited.contains(objId) || depth >= maxDepth) {
            return nodeBuilder.build();
        }
        
        visited.add(objId);
        
        // Extract methods
        List<MethodInfo> methods = extractMethods(obj);
        nodeBuilder.methods(methods);
        
        // Explore fields (sub-objects)
        for (Field field : clazz.getDeclaredFields()) {
            // Skip static fields
            if (Modifier.isStatic(field.getModifiers())) {
                continue;
            }
            
            // Skip private fields if not including private
            if (!includePrivate && !Modifier.isPublic(field.getModifiers())) {
                field.setAccessible(true); // Still try to access for dependency analysis
            }
            
            try {
                field.setAccessible(true);
                Object fieldValue = field.get(obj);
                
                if (fieldValue != null && isExplorableObject(fieldValue)) {
                    Set<Integer> visitedCopy = new HashSet<>(visited);
                    ObjectNode childNode = exploreObject(fieldValue, field.getName(), visitedCopy, depth + 1);
                    nodeBuilder.addChild(childNode);
                }
            } catch (IllegalAccessException e) {
                logger.fine("Could not access field: " + field.getName());
            }
        }
        
        return nodeBuilder.build();
    }

    /**
     * Determine if an object should be explored recursively.
     */
    private boolean isExplorableObject(Object obj) {
        if (obj == null) {
            return false;
        }
        
        Class<?> clazz = obj.getClass();
        
        // Skip primitives and wrappers
        if (clazz.isPrimitive()) {
            return false;
        }
        
        // Skip common types
        if (obj instanceof String || obj instanceof Number || obj instanceof Boolean ||
            obj instanceof Character || obj instanceof Enum) {
            return false;
        }
        
        // Skip collections (could be extended to explore their contents)
        if (obj instanceof Collection || obj instanceof Map) {
            return false;
        }
        
        // Skip arrays of primitives
        if (clazz.isArray() && clazz.getComponentType().isPrimitive()) {
            return false;
        }
        
        return true;
    }

    /**
     * Explore a class (without instantiation).
     * 
     * @param clazz The class to analyze
     * @return ObjectNode representing the class structure
     */
    public ObjectNode exploreClass(Class<?> clazz) {
        String className = clazz.getSimpleName();
        
        ObjectNode.Builder nodeBuilder = ObjectNode.builder(className, className)
                .depth(0);
        
        // Extract methods
        List<MethodInfo> methods = extractMethods(clazz);
        nodeBuilder.methods(methods);
        
        return nodeBuilder.build();
    }

    /**
     * Save object hierarchy as JSON file.
     * 
     * @param obj The object to analyze
     * @return Path to the saved file
     */
    public String saveAsJson(Object obj) throws IOException {
        return saveAsJson(obj, null);
    }

    /**
     * Save object hierarchy as JSON file with custom filename.
     * 
     * @param obj The object to analyze
     * @param filename Custom filename (null for auto-generated)
     * @return Path to the saved file
     */
    public String saveAsJson(Object obj, String filename) throws IOException {
        if (filename == null) {
            long timestamp = System.currentTimeMillis() / 1000;
            String className = obj.getClass().getSimpleName().toLowerCase();
            filename = className + "_" + timestamp + ".json";
        }
        
        String filepath = outputDir + File.separator + filename;
        
        ObjectNode rootNode = exploreObject(obj);
        
        try (FileWriter writer = new FileWriter(filepath)) {
            writer.write(rootNode.toJson());
        }
        
        logger.info("Saved JSON context to: " + filepath);
        return filepath;
    }

    /**
     * Save object hierarchy as AI-optimized text file.
     * 
     * @param obj The object to analyze
     * @return Path to the saved file
     */
    public String saveAsText(Object obj) throws IOException {
        return saveAsText(obj, null);
    }

    /**
     * Save object hierarchy as AI-optimized text file with custom filename.
     * 
     * @param obj The object to analyze
     * @param filename Custom filename (null for auto-generated)
     * @return Path to the saved file
     */
    public String saveAsText(Object obj, String filename) throws IOException {
        if (filename == null) {
            long timestamp = System.currentTimeMillis() / 1000;
            String className = obj.getClass().getSimpleName().toLowerCase();
            filename = className + "_" + timestamp + ".txt";
        }
        
        String filepath = outputDir + File.separator + filename;
        
        ObjectNode rootNode = exploreObject(obj);
        
        try (FileWriter writer = new FileWriter(filepath)) {
            writer.write(rootNode.toText());
        }
        
        logger.info("Saved text context to: " + filepath);
        return filepath;
    }

    /**
     * Get the output directory.
     */
    public String getOutputDir() {
        return outputDir;
    }

    /**
     * Get the maximum exploration depth.
     */
    public int getMaxDepth() {
        return maxDepth;
    }
}
